### PR TITLE
support restart container

### PR DESCRIPTION
**Pull request by chungngoops**
Related issue: #606 
GenericContainer:
-separate tryStart to createContainer & startContainer
-create public method restart()
ResourceReaper:
-separate stopAndRemoveContainer to stopContainer & removeContainer

### CODE REVIEW

**Comment by rnorth** [comment]
(Restarting failed builds just in case...)

**Comment by rnorth** [comment]
We have some merge conflicts - I'll resolve.

**Comment by kiview** [comment]
@rnorth I think the conflict is because we removed the Profiler, isn't it?

File: core/src/main/java/org/testcontainers/containers/DockerComposeContainer.java [hunk 1]
```diff
@@ -286,6 +286,11 @@ public void stop() {
         }
     }
 
+    @Override
+    public void restart(){
```
[the remark this answers is not in the document]
Comment (kiview) [reply core/src/main/java/org/testcontainers/containers/DockerComposeContainer.java]: We could throw an `UnsupportedException` here.

File: core/src/main/java/org/testcontainers/containers/GenericContainer.java [hunk 2]
```diff
@@ -326,6 +337,34 @@ private void connectToPortForwardingNetwork(String networkMode) {
...
+     * Restarts the container.
+     */
+    @Override
+    public void restart() {
+
+        if (containerId == null) {
+            return;
+        }
+
+        Profiler profiler = new Profiler("Container restart");
```
Comment (kiview) [inline core/src/main/java/org/testcontainers/containers/GenericContainer.java]: We can remove the `Profiler` now.

File: core/src/main/java/org/testcontainers/utility/ResourceReaper.java [hunk 3]
```diff
@@ -222,20 +233,26 @@ private void stopContainer(String containerId, String imageName) {
...
+                    "Error encountered when checking container for shutdown (ID: {}) - it may not have been stopped, or may already be stopped: {}",
+                    containerId, e.getMessage());
             return;
         }
 
         if (running) {
             try {
                 LOGGER.trace("Stopping container: {}", containerId);
-                dockerClient.killContainerCmd(containerId).exec();
+                dockerClient.stopContainerCmd(containerId).exec();
```
Comment (bsideup) [inline core/src/main/java/org/testcontainers/utility/ResourceReaper.java]: it is important to keep the `kill` behaviour here, why do you need to change it to `stop`?

File: core/src/main/java/org/testcontainers/lifecycle/Startable.java [hunk 4]
```diff
@@ -6,6 +6,8 @@
 
     void stop();
 
+    void restart();
```
Comment (bsideup) [inline core/src/main/java/org/testcontainers/lifecycle/Startable.java]: I would not introduce this method here. Also, it breaks the backward compatibility

File: core/src/main/java/org/testcontainers/containers/GenericContainer.java [hunk 5]
```diff
@@ -300,6 +312,31 @@ private void connectToPortForwardingNetwork(String networkMode) {
         });
     }
 
+    /**
+     * Restarts the container.
+     */
+    @Override
+    public void restart() {
+
```
Comment (bsideup) [inline core/src/main/java/org/testcontainers/containers/GenericContainer.java]:
```suggestion
```

**Comment by bsideup** [comment]
@chungngoops I did a quick test:
```
$ docker run -d -p 80 nginx
e8c8a526a5b173fe66b619d40e3b0bc2282b38d8ce8173b3efe7309c2128a947

$ docker ps -a
CONTAINER ID        IMAGE               COMMAND                  CREATED             STATUS              PORTS                   NAMES
e8c8a526a5b1        nginx               "nginx -g 'daemon of…"   5 seconds ago       Up 2 seconds        0.0.0.0:32768->80/tcp   priceless_mestorf

$ docker stop e8c8a526a5b1
e8c8a526a5b1

$ docker start e8c8a526a5b1
e8c8a526a5b1

$ docker ps -a
CONTAINER ID        IMAGE               COMMAND                  CREATED             STATUS              PORTS                   NAMES
e8c8a526a5b1        nginx               "nginx -g 'daemon of…"   14 seconds ago      Up 1 second         0.0.0.0:32769->80/tcp   priceless_mestorf
```

it seems that the ports are changed after the restart. It will break a lot of things, so I'm afraid we can't introduce this functionality in a generic way.

If you have a need, you can always use the client and call stop/start with the container id.

/cc @rnorth @kiview

File: core/src/main/java/org/testcontainers/utility/ResourceReaper.java (hunk 3 above)
Comment (chungngoops) [reply core/src/main/java/org/testcontainers/utility/ResourceReaper.java]: the purpose is to stop the container in a grateful way.

File: core/src/main/java/org/testcontainers/lifecycle/Startable.java (hunk 4 above)
Comment (kiview) [reply core/src/main/java/org/testcontainers/lifecycle/Startable.java]: You mean backward compatibility with possible user implementations of `Startable`?

**Comment by bsideup** [comment]
Closing due to the problem with ports. @chungngoops feel free to reopen/resubmit if you have ideas how that can be solved in a generic way

**Comment by carstenartur** [comment]
Is there any other way to use testcontainers for testing resilience regarding database hickups?
Would using a fixed port number in testcontainers be possible: ‘docker run -p 34567:8080 ubuntu bash’
see https://forums.docker.com/t/is-it-possible-to-keep-the-random-assigned-port-when-restarting-containers/25078/2

**Comment by bsideup** [comment]
The solution you linked is a hack that is vulnerable to the port conflicts and race conditions.

Consider using https://www.testcontainers.org/modules/toxiproxy/ or restarting the process **inside** the container (e.g. by using `execInContainer`)

**Comment by carstenartur** [comment]
Hi @bsideup,
thanks for sharing this info. No doubt the toxiproxy way is interesting and after some struggling with the configuration I finally got the toxiproxy to handle my mssql server connection using this approach up and running. However it does not do what I need as all pooled database connections are still valid after a temporary break on the database connection. So restarting the container or a process inside the container is more what I need to make sure that I need all database connections to be recreated.
Maybe this issue "support restart container" is in my situation overkill and maybe for all database images of testcontainers a way to restart the database within the docker image is much more suitable. So if you ask me if it would be possible to implement a common method to do the right execInContainer to do an internal restart of the database that would be a nice solution and an answer to your question on how it can be solved in an generic way at least for the database images.
Best regards,
Carsten

**Pull request closed without merging**